File: Problems/0200_0299/0208_Implement_Trie_(Prefix_Tree)/Project_Java/Trie.java
import java.util.*;

class TrieNode {
    // Initialize your data structure here.
    public TrieNode[] children;
    public boolean isWord;
    public TrieNode() {
        children = new TrieNode[26];    
    }
}

class Trie {
    // 29ms
    private TrieNode root;
    private boolean startWith;

    /** Initialize your data structure here. */
    public Trie() {
        root = new TrieNode();
    }
    
    /** Inserts a word into the trie. */
    public void insert(String word) {
        insert(word, root, 0);
    }
    
    private void insert(String word, TrieNode root, int idx) {
        if (idx == word.length()) {
            root.isWord = true;
            return;
        }

        int index = word.charAt(idx) - 'a';

        if (root.children[index] == null) {
            root.children[index] = new TrieNode();
        }

        insert(word, root.children[index], idx + 1);         
    }
    
    /** Returns if the word is in the trie. */
    public boolean search(String word) {
        return search(word, root, 0);         
    }

    public boolean search(String word, TrieNode root, int idx) {
        if (idx == word.length()) {
            startWith = true;
            return root.isWord;
        }

        int index = word.charAt(idx) - 'a';

        if (root.children[index] == null) {
            startWith = false;
            return false;
        }
            
        return  search(word,root.children[index], idx + 1);
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(String prefix) {
        search(prefix);
        return startWith;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */

File: Problems/0200_0299/0208_Implement_Trie_(Prefix_Tree)/Project_Go/solution/trie_bad/trie.go
package trie

import "./trienode"

type Trie struct {
	root      trienode.TrieNode
	startWith bool
}

/** Initialize your data structure here. */
func Constructor() Trie {
	trie := new(Trie)
	trie.root = trienode.Constructor()
	return *trie
}

/** Inserts a word into the trie. */
func (this *Trie) Insert(word string) {
	insert2(word, this.root, 0)
}

func insert2(word string, node trienode.TrieNode, idx int) {
	if idx == len(word) {
		node.IsWord = true
		return
	}

	index := word[idx] - 'a'

	//	if node.Children[index] == nil {
	if node.IsExistChildren[index] == false {
		node.Children[index] = trienode.Constructor()
		node.IsExistChildren[index] = true
	}

	insert2(word, node.Children[index], idx+1)
}

/** Returns if the word is in the trie. */
func (this *Trie) Search(word string) bool {
	return this.Search2(word, this.root, 0)
}

func (this *Trie) Search2(word string, node trienode.TrieNode, idx int) bool {
	if idx == len(word) {
		this.startWith = true
		return node.IsWord
	}

	index := word[idx] - 'a'

	if node.IsExistChildren == nil {
		this.startWith = false
		return false
	}

	if node.IsExistChildren[index] == false {
		this.startWith = false
		return false
	}

	return this.Search2(word, this.root.Children[index], idx+1)
}

/** Returns if there is any word in the trie that starts with the given prefix. */
func (this *Trie) StartsWith(prefix string) bool {
	this.Search(prefix)
	return this.startWith
}

/**
 * Your Trie object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Insert(word);
 * param_2 := obj.Search(word);
 * param_3 := obj.StartsWith(prefix);
 */
